### ISSUE TITLE
`CanBeNonNullable` false positive 

### ISSUE THREAD

**Issue by eygraber** [detekt/detekt#5629]
This is happening in 1.22.0 and before as well https://github.com/detekt/detekt/issues/4677

I copied it from my project and sanitized it, so I'm not sure what the minimum repro is here.

The error is:

```
CanBeNonNullable - [The nullable parameter 'bar' can be made non-nullable.] at Playground.kt:107:33
```

And the repro is:

```
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlin.random.Random

private suspend inline fun log(msg: () -> String) {
  delay(1_000)
  println(msg())
}

const val SHARED_FLOW_DEFAULT_BUFFER = 64

fun <T> bufferedSharedFlow(
  capacity: Int = SHARED_FLOW_DEFAULT_BUFFER,
  onBufferOverflow: BufferOverflow = BufferOverflow.SUSPEND
) =
  MutableSharedFlow<T>(extraBufferCapacity = capacity, onBufferOverflow = onBufferOverflow)

sealed class BarType {
  object One : BarType()
  object Two : BarType()
}

class Bar {
  val type
    get() = when(System.currentTimeMillis()) {
      0L -> BarType.One
      else -> BarType.Two
    }
}

internal interface BarHelper {
  suspend fun findNextBar(targetBar: Bar? = null): Bar?
}

internal class BarHelperImpl : BarHelper {
  override suspend fun findNextBar(targetBar: Bar?): Bar? {
    val bars = listOf(Bar(), Bar())
    val nextBar =
      bars
        .find { bar ->
          bar.type is BarType.One && when(val type = bar.type) {
            is BarType.One -> withContext(Dispatchers.IO) {
              type == BarType.One
            }

            else -> true
          }
        }

    val isRandomTrue = Random.nextBoolean()

    return when {
      isRandomTrue -> nextBar

      else -> targetBar ?: nextBar
    }
  }
}

internal interface Foo {
  suspend fun foo(bar: Bar? = null): Bar?
}

internal interface Baz {
  fun baz(): Flow<Bar?>
}

internal class BazImpl : Baz {
  private val b = bufferedSharedFlow<Bar?>(capacity = 1)

  override fun baz() = b
}

internal interface Bang

internal class BangImpl(
  private val baz: Baz,
  private val barHelper: BarHelper,
  private val scope: CoroutineScope
) {
  fun start() {
    with(scope) {
      launch {
        bang()
      }
    }
  }

  private fun CoroutineScope.bang() {
    launch {
      baz
        .baz()
        .collect { bar ->
          log { "Bang (bar=$bar)" }

          handleBar(barHelper.findNextBar(bar))
        }
    }
  }

  private suspend fun handleBar(bar: Bar?) { // error is on this line
    when(bar?.type) {
      BarType.One -> log { "One" }
      BarType.Two -> log { "Two" }
      null -> log { "Null" }
    }
  }
}
```

**Comment by atulgpt** [comment]
Hi @eygraber/ @BraisGabin  the minimal reproducible failing TC as per your code will be
```kotlin
@Test
fun test() {
    val code = """
        import kotlinx.coroutines.withContext
        import java.util.UUID
        import kotlin.random.Random

        fun log(msg: String) {
          println(msg)
        }

        class RandomBar {
          val id
            get() = UUID.randomUUID().toString()
        }

        class BangImpl {
          private fun handleBar(bar: RandomBar?) { // error is on this line
            when(bar?.id) {
              "1" -> log("One")
              null -> log("Null")
            }
          }
        }
    """.trimIndent()
    assertThat(subject.compileAndLintWithContext(env, code)).isEmpty()
}
```

But my question is why this is marked as a false positive as the `CanBeNonNullable` suggest which value parameter can be non-null. In your example complaint code would be like below
```kotlin
private suspend inline fun log(msg: () -> String) {
    delay(1_000)
    println(msg())
}

const val SHARED_FLOW_DEFAULT_BUFFER = 64

fun <T> bufferedSharedFlow(
    capacity: Int = SHARED_FLOW_DEFAULT_BUFFER,
    onBufferOverflow: BufferOverflow = BufferOverflow.SUSPEND
) =
    MutableSharedFlow<T>(extraBufferCapacity = capacity, onBufferOverflow = onBufferOverflow)

sealed class BarType {
    object One : BarType()
    object Two : BarType()
}

class Bar {
    val type
        get() = when(System.currentTimeMillis()) {
            0L -> BarType.One
            else -> BarType.Two
        }
}

internal interface BarHelper {
    suspend fun findNextBar(targetBar: Bar? = null): Bar?
}

internal class BarHelperImpl : BarHelper {
    override suspend fun findNextBar(targetBar: Bar?): Bar? {
        val bars = listOf(Bar(), Bar())
        val nextBar =
            bars
                .find { bar ->
                    bar.type is BarType.One && when(val type = bar.type) {
                        is BarType.One -> withContext(Dispatchers.IO) {
                            type == BarType.One
                        }

                        else -> true
                    }
                }

        val isRandomTrue = Random.nextBoolean()

        return when {
            isRandomTrue -> nextBar

            else -> targetBar ?: nextBar
        }
    }
}

internal interface Foo {
    suspend fun foo(bar: Bar? = null): Bar?
}

internal interface Baz {
    fun baz(): Flow<Bar?>
}

internal class BazImpl : Baz {
    private val b = bufferedSharedFlow<Bar?>(capacity = 1)

    override fun baz() = b
}

internal interface Bang

internal class BangImpl(
    private val baz: Baz,
    private val barHelper: BarHelper,
    private val scope: CoroutineScope
) {
    fun start() {
        with(scope) {
            launch {
                bang()
            }
        }
    }

    private fun CoroutineScope.bang() {
        launch {
            baz
                .baz()
                .collect { bar ->
                    log { "Bang (bar=$bar)" }

                    barHelper.findNextBar(bar)?.let {
                        handleBar(it)
                    } ?: run {
                        "Null"
                    }
                }
        }
    }

    private suspend fun handleBar(bar: Bar) { // error is on this line
        when(bar.type) {
            BarType.One -> log { "One" }
            BarType.Two -> log { "Two" }
        }
    }
}

```

**Comment by eygraber** [comment]
I think it's a false positive because the function can receive a null argument. Rewriting the function isn't the correct move here; what if there were multiple callers of `handleBar`? They shouldn't all have to hoist the logging out.

**Comment by atulgpt** [comment]
Hi, @eygraber spoke too soon 😅. Now I understand the rule and yes it is a false positive. @BraisGabin I think I can pick this up if it is available

**Comment by BraisGabin** [comment]
Go for it!

**Comment by atulgpt** [comment]
Hi @BraisGabin, I have added a few failing TCs at https://github.com/detekt/detekt/pull/5714  Can you review those to validate my understanding?

**Comment by eygraber** [comment]
Not sure if this is a separate issue or not, but got another false positive for `CanBeNonNullable`.

If my code looks like this, I get the false positive:

```kotlin
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.consumeWindowInsets
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier

data class ScaffoldState(
  val appBarTitle: String? = null
)

@OptIn(ExperimentalLayoutApi::class, ExperimentalMaterial3Api::class)
@Composable
fun DetektReproScaffold(
  state: ScaffoldState,
  shouldNavigate: Boolean,
  modifier: Modifier = Modifier
) {
  Scaffold(
    modifier = modifier,
    topBar = {
      HtTopAppBar(
        shouldNavigate = shouldNavigate,
        title = state.appBarTitle
      )
    }
  ) { innerPadding ->
    Box(
      modifier = Modifier
        .consumeWindowInsets(innerPadding)
        .padding(innerPadding)
    )
  }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun HtTopAppBar(
  shouldNavigate: Boolean,
  modifier: Modifier = Modifier,
  title: String? = null
) {
  if(shouldNavigate || title != null) {
    TopAppBar(
      title = {
        if(title != null) {
          Text(
            text = title
          )
        }
      },
      modifier = modifier
    )
  }
}
```

However, if I move the `if` outside of the function then the false positive does not occur:

```kotlin
@OptIn(ExperimentalLayoutApi::class, ExperimentalMaterial3Api::class)
@Composable
fun DetektReproScaffold(
  state: ScaffoldState,
  shouldNavigate: Boolean,
  modifier: Modifier = Modifier
) {
  Scaffold(
    modifier = modifier,
    topBar = {
      if(shouldNavigate || state.appBarTitle != null) {
        HtTopAppBar(
          title = state.appBarTitle
        )
      }
    }
  ) { innerPadding ->
    Box(
      modifier = Modifier
        .consumeWindowInsets(innerPadding)
        .padding(innerPadding)
    )
  }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun HtTopAppBar(
  modifier: Modifier = Modifier,
  title: String? = null
) {
  TopAppBar(
    title = {
      if(title != null) {
        Text(
          text = title
        )
      }
    },
    modifier = modifier
  )
}
```

**Comment by BraisGabin** [comment]
This is a different issue. I think that the rule doesn't differentiate between `if(shouldNavigate || title != null)` and `if(shouldNavigate && title != null)`.

The first one shouldn't be flagged but the second should.

**Comment by eygraber** [comment]
Sure

**Comment by eygraber** [comment]
https://github.com/detekt/detekt/issues/5799
